### PR TITLE
Reduce TrieNode allocations

### PR DESCRIPTION
**Pull request by Scooletz**
This PR removes 10% managed allocations from the node. It does it by moving `StorageRoot` from `TrieNode`  as an item of an array and getting rid of `RlpStream` from `TrieNode` and delegating its functionality to `Rlp.ValueDecoderContext`

### Benchmarks

Reused `PatriciaTreeBenchmarks` that provides a coarse grained measurement of various operations.

```bash
BenchmarkDotNet=v0.13.5, OS=Windows 11 (10.0.22621.1413/22H2/2022Update/SunValley2)
12th Gen Intel Core i9-12900HK, 1 CPU, 20 logical and 14 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
  DefaultJob : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX2
```

|   Method |     Mean |   Error |  StdDev |  Median|    Gen0 |   Gen1 | Allocated |
|--------- |---------:|--------:|--------:|--------:|-------:|----------:|----------:|
| `master` | 245.4 us | 8.13 us | 23.84 us | 254.4 us | 12.4512 | 3.1738 | 155.43 KB |
| `this PR` | 181.5 us | 9.02 us | 26.59 us | 193.7 us | 10.2539 | 2.6855 | 126.38 KB |

### Running node

| Version | Memory allocation - min| Memory allocation - max |  Memory allocation -  avg | New Payload Execution Time - Mean | New Payload Execution Time - Stdev |
|--------- |---------:|--------:|--------:|--------:|-------:|
| `master` | 4.11 MB/s | 21.8 MB/s |  5.46 MB/s | 891 | 496  |
| `this PR` | 3.90 MB/s|  12.5 MB/s | 4.80 MB/s | 779 | 408 |

<details>
<summary>**Image** comparing of all the metrics </summary>

image

</details>

## Changes

- `StorageRoot` is now stored as `_data[2]` only in leafs
- `RlpStreamReader` used instead of the object of `RlpStream`

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [x] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

## Remarks

Another approach to the `TrieNode` optimization.

### CODE REVIEW

**Comment by Scooletz** [comment]
Maybe I can change the implementation to use `Rlp.ValueDecoderContext`?

File: src/Nethermind/Nethermind.Serialization.Rlp/RlpStreamReader.cs [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+        public RlpStreamReader(byte[] data) => Data = data;
+
+        private string Description =>
+            Data?.Slice(0, Math.Min(Rlp.DebugMessageContentLength, Length)).ToHexString() ?? "0x";
+
+        public bool IsNull => Data is null;
+
+        public byte[]? Data { get; }
+
+        public int Position { get; set; }
```
[the remark this answers is not in the document]
Comment (Scooletz) [reply src/Nethermind/Nethermind.Serialization.Rlp/RlpStreamReader.cs:34]: I think I'll leave as is. The stack is the stack one int or byte wont' make that much change. Also, I'll default to `Rlp.ValueContext` for now.

**Review by LukaszRozmej — APPROVED** [review]

**Comment by Scooletz** [comment]
Doing a full sync on teku mainnet. Once it's synced and working will confirm and merge if the results are consistent.

**Comment by Scooletz** [comment]
The teku mainnet has been synced with this and now is operational. Waiting for a few hours of running and will be merging if nothing contradicts the findings above.

**Comment by Scooletz** [comment]
I confirm that with the usage of the Rlp context reading the version behaves as it was with the copied `RlpStreamReader`, reducing the measured alloc ratio by at least 10%. Merging.

**Pull request merged**